pennylane_lightning/lightning_gpu/lightning_gpu.py

```diff
@@ -850,14 +847,18 @@ def processing_fn(tape):
         # pylint: disable=attribute-defined-outside-init
         def sample(self, observable, shot_range=None, bin_size=None, counts=False):
             """Return samples of an observable."""
-            if observable.name != "PauliZ":
-                self.apply_lightning(observable.diagonalizing_gates())
+            diagonalizing_gates = observable.diagonalizing_gates()
+            if diagonalizing_gates:
+                self.apply(diagonalizing_gates)
+            if not isinstance(observable, qml.PauliZ):
                 self._samples = self.generate_samples()
-            return super().sample(
+            results = super().sample(
                 observable, shot_range=shot_range, bin_size=bin_size, counts=counts
             )
+            if diagonalizing_gates:
+                self.apply([qml.adjoint(g, lazy=False) for g in reversed(diagonalizing_gates)])
```

Comment:
why is this lazy set to `False` required?

Reply:
@albi3ro suggested this change. The doc says 

> If lazy=False, operation-specific adjoint decompositions are first attempted.

which makes it more likely to give another "native" operation, instead of a nested operation class. So I think in essence this forces, for example, `qml.Adjoint(qml.RX(0.1))` to `qml.RX(-0.1)` instead of something fancier.